core/src/erasure.rs

```diff
@@ -470,6 +477,78 @@ pub fn recover(
     Ok((missing_data, missing_coding))
 }
 
+pub fn recover_with_blobs(
+    slot: u64,
+    start_idx: u64,
+    blobs_to_recover: &[Option<Vec<u8>>; NUM_CODING + NUM_DATA],
+) -> Result<(Vec<SharedBlob>, Vec<SharedBlob>)> {
+    let block_start_idx = start_idx - (start_idx % NUM_DATA as u64);
+
+    debug!("block_start_idx: {}", block_start_idx);
+
+    let coding_start_idx = block_start_idx + NUM_DATA as u64 - NUM_CODING as u64;
+    let block_end_idx = block_start_idx + NUM_DATA as u64;
+    trace!(
+        "recover: coding_start_idx: {} block_end_idx: {}",
+        coding_start_idx,
+        block_end_idx
+    );
+
+    let mut blobs: Vec<SharedBlob> = Vec::with_capacity(ERASURE_SET_SIZE);
+    let mut erasures: Vec<i32> = Vec::with_capacity(NUM_CODING);
+
+    let mut missing_data: Vec<SharedBlob> = vec![];
+    let mut missing_coding: Vec<SharedBlob> = vec![];
+
+    if blobs_to_recover[NUM_DATA..].iter().all(Option::is_none) {
+        return Err(Error::ErasureError(ErasureError::NotEnoughBlocksToDecode));
+    }
+
+    let mut size = None;
+
+    for (erasure_index, blob_bytes_opt) in blobs_to_recover.iter().enumerate() {
+        match blob_bytes_opt {
+            Some(bytes) => {
+                if bytes.len() <= BLOB_HEADER_SIZE || bytes.len() > BLOB_SIZE {
+                    return Err(Error::ErasureError(ErasureError::InvalidBlobData));
+                }
+                if erasure_index as usize >= NUM_DATA && size.is_none() {
+                    size = Some(bytes.len() - BLOB_HEADER_SIZE);
+                }
+                blobs.push(Arc::new(RwLock::new(Blob::new(&bytes))));
```

Comment:
would be really nice to avoid putting blobs, then rwlock, then arc on top of the vectors only to have them stripped off again in decode_blobs() ...